akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -210,6 +211,178 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
 /**
  * INTERNAL API
  */
+final class GroupBy[T, K](maxSubstreams: Int, keyFor: T ⇒ K) extends GraphStage[FlowShape[T, Source[T, NotUsed]]] {
+  val in: Inlet[T] = Inlet("GroupBy.in")
+  val out: Outlet[Source[T, NotUsed]] = Outlet("GroupBy.out")
+  override val shape: FlowShape[T, Source[T, NotUsed]] = FlowShape(in, out)
+  override def initialAttributes = DefaultAttributes.groupBy
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) with OutHandler with InHandler {
+    parent ⇒
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    private val activeSubstreamsMap = new java.util.HashMap[Any, SubstreamSource]()
+    private val closedSubstreams = new java.util.HashSet[Any]()
+    private var timeout: FiniteDuration = _
+    private var substreamWaitingToBePushed: Option[SubstreamSource] = None
+    private var nextElementKey: K = null.asInstanceOf[K]
+    private var nextElementValue: T = null.asInstanceOf[T]
+    private var _nextId = 0
+    private val substreamsJustStared = new java.util.HashSet[Any]()
+    private var firstPushCounter: Int = 0
+
+    private def nextId(): Long = { _nextId += 1; _nextId }
+
+    private def hasNextElement = nextElementKey != null
+
+    private def clearNextElement(): Unit = {
+      nextElementKey = null.asInstanceOf[K]
+      nextElementValue = null.asInstanceOf[T]
+    }
+
+    private def tryCompleteAll(): Boolean =
+      if (activeSubstreamsMap.isEmpty || (!hasNextElement && firstPushCounter == 0)) {
+        for (value ← activeSubstreamsMap.values()) value.complete()
+        completeStage()
+        true
+      } else false
+
+    private def fail(ex: Throwable): Unit = {
+      for (value ← activeSubstreamsMap.values()) value.fail(ex)
+      failStage(ex)
+    }
+
+    private def needToPull: Boolean = !(hasBeenPulled(in) || isClosed(in) || hasNextElement)
+
+    override def preStart(): Unit = timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
+
+    override def onPull(): Unit = {
+      substreamWaitingToBePushed match {
+        case Some(substreamSource) ⇒
+          push(out, Source.fromGraph(substreamSource.source))
+          scheduleOnce(substreamSource.key, timeout)
+          substreamWaitingToBePushed = None
+        case None ⇒
+          if (hasNextElement) {
+            val subSubstreamSource = activeSubstreamsMap.get(nextElementKey)
+            if (subSubstreamSource.isAvailable) {
```

Comment:
Back to conversation that @johanandren  started:
>Not only theoretically: 
1. push from upstream, elem with new key k1
2. pull from downstream
3. emit new substream for k1
4. pull from substream k1 consuming firstElement
5. push from upstream with key k1, before demand from substream k1 - boom

I was thinking about this.
So this for onPull method for main stream.  To have it called one out of two must be true:
1) Stream just started - we can skip this case
2) GroupBy just pushed new substream
To push new substream we need as previous step one of following:
- have pull from one of substreams - and this is not possible when nextElement is there (needToPull method)
- have pull from downstream - this means that groupBy just pushed new substream or stream just started

Even if push to substream is long enough and subSubstreamSource.isAvailable = false within several push/pull cycles - pull to upstream will not be sent until nextElement has some element 

CC: @drewhk